lib/ecto/adapters/sql.ex

```diff
@@ -458,43 +449,85 @@ defmodule Ecto.Adapters.SQL do
     end
   end
 
+  defp do_transaction(repo, opts, fun) do
+    {pool_mod, pool, timeout} = repo.__pool__
+    opts    = Keyword.put_new(opts, :timeout, timeout)
+    timeout = Keyword.fetch!(opts, :timeout)
+
+    case Pool.open_transaction(pool_mod, pool, timeout) do
+      {:ok, ref, {mod, _conn}, queue_time} ->
+        try do
+          mode = transaction_mode(pool_mod, pool, timeout)
+          transaction(repo, ref, mod, mode, queue_time, timeout, opts, fun)
+        catch
+          # If any error leaked, it was inside Ecto.
+          # Make sure we break the connection.
+          kind, reason ->
+            stack = System.stacktrace()
+            Pool.break(ref)
```

Comment:
It doesn't matter because close_transaction is no-op if it was broken. But I guess else is better semantically, what do you think?